File: java/src/main/java/com/fish/list/Q142.java
package com.fish.list;

/**
 * @author: liuqi
 * @date: 2024/3/6
 */
public class Q142 {
    /**
     * 这种题先画图，理解数学原理
     */
    public class Solution {
        public ListNode detectCycle(ListNode head) {
            ListNode slow = head;
            ListNode fast = head;
            while (fast != null && fast.next != null) {
                slow = slow.next;
                fast = fast.next.next;
                if (fast == slow) {
                    while (slow != head) {
                        slow = slow.next;
                        head = head.next;
                    }
                    return head;
                }
            }
            return null;
        }
    }
}

File: C++/Q142.cpp
struct ListNode
{
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(nullptr) {}
};
class Solution
{
public:
    ListNode *detectCycle(ListNode *head)
    {
        ListNode *fast = head;
        ListNode *slow = head;
        while (fast)
        {
            slow = slow->next;
            if (!fast->next)
            {
                return nullptr;
            }
            fast = fast->next->next;
            if (fast == slow)
            {
                fast = head;
                while (fast != slow)
                {
                    fast = fast->next;
                    slow = slow->next;
                }
                return fast;
            }
        }
        return nullptr;
    }
};